docs: refresh codec server page (#4483)
temporalio/documentation: docs/encyclopedia/data-conversion/codec-server.mdx

id: codec-server
title: Codec Server
sidebar_label: Codec Server
description:
  A Codec Server is an HTTP server that provides remote encoding and decoding for Temporal Payloads, enabling the Web UI
  and CLI to display decoded data without exposing encryption keys to the Temporal Service.
slug: /codec-server
toc_max_heading_level: 4
keywords:
  - encryption
  - codec-server
  - payloads
  - data-converters

>   - codec-server

Author: ```suggestion
```

>   - remote-data-encoding

Author: ```suggestion
  - codec-server
```

To support External Storage, create a handler using `NewPayloadHTTPHandler` with `PayloadHTTPHandlerOptions`. The options
accept your storage drivers, your pre-storage codecs (the Payload Codecs configured in your Worker's Data Converter),
and any post-storage codecs (codecs applied by a proxy after external storage). The handler applies them in the correct
order across all endpoints automatically. When you configure the handler with storage drivers, the existing endpoints
become storage-aware and a new `/download` endpoint becomes available:

:::caution

`NewPayloadHTTPHandler` runs the full encode-store-encode and decode-retrieve-decode pipeline. Do not use it as a target
for a remote Data Converter or remote codec on your Workers. For remote codecs, use `NewPayloadCodecHTTPHandler`
separately. If you need both, set up `NewPayloadHTTPHandler` for the Web UI and CLI alongside
`NewPayloadCodecHTTPHandler` for your Workers, and configure both with the same codecs.

:::

- **`/download`** retrieves the actual payload data from external storage and decodes it through the Payload Codec. This

>   endpoint is used internally by `/decode` when it encounters storage references, but you can also call it directly from

Reviewer: Small typo?: "call it directly from to" should probably be "call it directly from **the Web UI** to", IIUC.

>   endpoint is used internally by `/decode` when it encounters storage references, but you can also call it directly from

Author: ```suggestion
  endpoint is used internally by `/decode` when it encounters storage references, but you can also call it directly from the Web UI 
```

independent of the Temporal Service. Encryption keys and codec logic remain in your environment.

For setup instructions, see [Codec Server setup](/production-deployment/data-encryption#codec-server-setup).

## Why use a Codec Server

When you apply a custom [Payload Codec](/payload-codec) for encryption or compression, data stored in the Temporal
Service is encoded. The Temporal Service never has access to your encryption keys, so it cannot decode this data.
Without a Codec Server, the Web UI and CLI display raw encoded payloads.

A Codec Server solves this by giving the Web UI and CLI a way to decode payloads on demand, without exposing keys to the
Temporal Service. Common reasons to run a Codec Server include:

- **Debugging Workflows.** View decoded Workflow inputs, outputs, and Event History in the Web UI instead of reading
  base64-encoded or encrypted blobs.
- **Operating from the CLI.** Use commands like `temporal workflow show` and `temporal workflow execute` with readable
  data, even when payloads are encrypted at rest.
- **Encoding inputs from the UI and CLI.** When you start or signal a Workflow from the Web UI or CLI, the Codec Server

>   can encode the input before it reaches the Temporal Service, so sensitive data is never sent in plaintext.

Reviewer: Small nit: "never sent in plaintext" is *partially* true, relative to the Temporal Service, but plaintext does cross from the user's browser/CLI to the Codec Server, ideally over TLS:

```suggestion
  can encode the input before it reaches the Temporal Service, so the Temporal Service never sees 
  plaintext (the input still travels from your browser or CLI to the Codec Server, which is why HTTPS 
  matters  in any non-loopback deployment).
```

exposes two HTTP POST endpoints:

- **`/encode`** accepts plaintext payloads and returns encoded payloads. Used for sending payloads.
- **`/decode`** accepts encoded payloads and returns decoded payloads. Used for retrieving payloads.

Both endpoints receive and respond with a JSON body containing a `payloads` array of [Payload](/dataconversion#payload)
objects. The Codec Server passes each payload through your [Payload Codec](/payload-codec), which applies the same
encoding or decoding logic that your Workers use.

<CaptionedImage
  src="/diagrams/codec-server.svg"
  srcDark="/diagrams/codec-server-dark.svg"
  width="100%"
  title="Codec Server"
/>

When the Web UI or CLI needs to display decoded data, it sends the encoded payloads to your Codec Server's `/decode`
endpoint. The Codec Server decodes the payloads and returns them to the client. The Temporal Service never sees the
decoded data.

The `/encode` endpoint works in the other direction. When you start a Workflow or send a Signal from the Web UI or CLI,
the input is sent to the Codec Server's `/encode` endpoint first, so data reaches the Temporal Service in its encoded
form.

> Your Codec Server should use the same Payload Codec implementation as your Workers to ensure consistent encoding and
> decoding.

Reviewer: ```suggestion
decoding. Because the Codec Server often holds the same keys as your Workers, treat its
host with the same trust as a Worker; anyone who can call it has effective decrypt parity.
```

> ## Codec Server vs. Payload Codec

Reviewer: A few points here we could consider clarifying:

1. HTTPS should be required for any non-loopback deployment.
2. JWT verification != authorization: verifying the JWT signature only proves the token came from your IdP or Temporal Cloud. Ideally, the codec server would _additionally check_ the  token's claims authorize the namespace in `X-Namespace`. This generally requires additional API calls today to be made from the Codec Server, since it's not encoded in the JWT itself, since the Auth0 minted JWT for Temporal Cloud is used to access more than one namespace in a web UI session.
3. Auth guidance: the "better" option is actually "Include cross-origin credentials", **_instead of_** "Pass the user access token". With it, customer codec server developers could use an existing organization-wide authentication provider (via the user browser’s cookies for an IdP) instead of the Auth0-minted JWT for Temporal Cloud. There's lots of nuance here, maybe we include diagrams for the two auth options?

- **`/decode`** accepts encoded payloads and returns decoded payloads.

Both endpoints receive and respond with a JSON body containing a `payloads` array of
[Payload](/dataconversion#payload) objects. The Codec Server passes each payload through your
[Payload Codec](/payload-codec), which applies the same encoding or decoding logic that your Workers use.

<CaptionedImage
  src="/diagrams/tctl-diagram-codec-server.svg"
  title="Codec Server"
  width="50%"
  zoom="true"
/>

When the Web UI or CLI needs to display decoded data, it sends the encoded payloads to your Codec Server's `/decode`
endpoint. The Codec Server decodes the payloads and returns them to the client. Decoding happens entirely on the client
side. The Temporal Service never sees the decoded data.

The `/encode` endpoint works in the other direction. When you start a Workflow or send a Signal from the Web UI or CLI,
the input is sent to the Codec Server's `/encode` endpoint first, so data reaches the Temporal Service in its encoded
form.

Your Codec Server should use the same Payload Codec implementation as your Workers to ensure consistent encoding and

> decoding.

Author: After this we can have a table that lists differences between a Codec Server and a Payload Codec. They both have the word Codec and it may be easy to confuse the two

> ## Web UI and CLI integration

Author: Most of this is how-to territory. Don't need this here to duplicate the content on the code server setup page

Temporal Service. Common reasons to run a Codec Server include:

- **Debugging Workflows.** View decoded Workflow inputs, outputs, and Event History in the Web UI instead of reading
  base64-encoded or encrypted blobs.
- **Operating from the CLI.** Use commands like `temporal workflow show` and `temporal workflow execute` with readable
  data, even when payloads are encrypted at rest.
- **Encoding inputs from the UI and CLI.** When you start or signal a Workflow from the Web UI or CLI, the Codec Server
  can encode the input before it reaches the Temporal Service, so sensitive data is never sent in plaintext.
- **Compliance and access control.** Because the Codec Server runs in your environment, you control who can decode
  payloads and under what conditions. You can layer authorization on top of the decode endpoint to restrict access per
  user or per Namespace.

## How a Codec Server works

A Codec Server follows the Temporal
[Codec Server Protocol](https://github.com/temporalio/samples-go/tree/main/codec-server#codec-server-protocol). It
exposes two HTTP POST endpoints:

> - **`/encode`** accepts plaintext payloads and returns encoded payloads.

Reviewer 2: ```suggestion
- **`/encode`** accepts plaintext payloads and returns encoded payloads.  Used for sending payloads.
```

> - **`/decode`** accepts encoded payloads and returns decoded payloads.

Reviewer 2: ```suggestion
- **`/decode`** accepts encoded payloads and returns decoded payloads.  Used for retrieving payloads.
```

Both endpoints receive and respond with a JSON body containing a `payloads` array of [Payload](/dataconversion#payload)
objects. The Codec Server passes each payload through your [Payload Codec](/payload-codec), which applies the same
encoding or decoding logic that your Workers use.

<CaptionedImage
  src="/diagrams/codec-server.svg"
  srcDark="/diagrams/codec-server-dark.svg"
  width="100%"
  title="Codec Server"
/>

When the Web UI or CLI needs to display decoded data, it sends the encoded payloads to your Codec Server's `/decode`
endpoint. The Codec Server decodes the payloads and returns them to the client. Decoding happens entirely on the client

> side. The Temporal Service never sees the decoded data.

Reviewer 2: ```suggestion
When the Web UI or CLI needs to display decoded data, it sends the encoded payloads to your Codec Server's `/decode`
endpoint. The Codec Server decodes the payloads and returns them to the client.  The Temporal Service never sees the decoded data.
```

Eliminating the seeming contradiction.  "codec server decodes ... decoding happens on the client"

| **Purpose**                       | Encodes and decodes Payloads. Applies encryption, compression, or other byte-level transformations. | Hosts a Payload Codec as an HTTP service so the Web UI and CLI can encode and decode Payloads remotely. |
| **Runs where**                    | In-process, inside your Workers and Clients. Also runs inside the Codec Server.                     | As a standalone HTTP service in your environment, with a Payload Codec inside it.                       |
| **Called by**                     | The Temporal SDK, automatically on every serialization and deserialization.                         | The Web UI and CLI, over HTTP, when a user views or submits Payload data.                               |
| **Has access to encryption keys** | Yes. Keys are available in the Worker or Client process.                                            | Yes. Must be configured with the same keys the Payload Codec uses.                                      |

You implement the transformation logic once in a Payload Codec, then host that logic in a Codec Server so the Web UI and
CLI can use it remotely.

> ## Authorization

Reviewer 2: This looks good but somebody from security, perhaps Kent Gruber, should have a look.

> Access to a Codec Server should be restricted because it can decode sensitive data with a single API call. Common

Reviewer 2: ```suggestion
Access to a Codec Server should be restricted because it can decode sensitive data. Common
```
